### PR TITLE
Add API to allow addons to define custom transform

### PR DESCRIPTION
**Pull request by kratiahuja**
[rendered](https://github.com/kratiahuja/rfcs/blob/add-import-api/active/0000-add-custom-transform.md)

cc: @stefanpenner @rwjblue

### CODE REVIEW

**Review by trentmwillis — COMMENTED** [review]
Overall, I like this and think it adds a nice bit of extensibility to the `import` API. That said, I think we should use a more specific name than "custom transformation".

Since the transform is only applicable for imports, we should likely refer to them as "import transforms" and the corresponding hook could then be `addImportTransform`. Otherwise, it will likely communicate that these transforms can be run on any code in the build pipeline, which I can see causing confusion.

**Comment by kratiahuja** [comment]
Agreed. I am terribly bad at naming things :)

**Review by hjdivad — COMMENTED** [review]
I agree with @trentmwillis although I'd also add that the [addon API](https://ember-cli.com/api/classes/Addon.html) uses nouns so it's a bit strange to add a function that starts with `add` but seems to function similarly to existing APIs.

My 2c on the bikeshedding would be for something like `importTransforms`.

---

Minor bikeshedding aside, what behaviour do we want for addons that provide transforms of the same name?  Shall we make this an error case initially?

**Comment by trentmwillis** [comment]
> What behaviour do we want for addons that provide transforms of the same name? Shall we make this an error case initially?

My opinion is that it should either error or, at the least, log a message stating which one is being used. We may also want to consider having an "override" functionality, in case a consumer wants to override a transform that is bundled together with something else (e.g., override the fastboot transform, but still needs the rest of the fastboot addon).

**Comment by hjdivad** [comment]
> My opinion is that it should either error or, at the least, log a message stating which one is being used. We may also want to consider having an "override" functionality, in case a consumer wants to override a transform that is bundled together with something else (e.g., override the fastboot transform, but still needs the rest of the fastboot addon).

Override with a warning is fine as long as we guarantee the order in which addons are processed for custom transforms.

**Comment by kratiahuja** [comment]
For the override functionality, I would actually prefer creating a `from` optional keyword which allows consumers of the transform to define which one they chose to pick. Example an addon `my-addon` provides a slightly different of `fastboot-shim` from `ember-cli-fastboot` and names it `fastboot-shim`:

```js
app.import('path/to/module.js', {
  'using': [{ transformation: 'fastboot-shim', from: 'my-addon'}]
});
```

If an addon importing the third party library choses to not use `from` keyword, then the last definition of `fastboot-shim` wins (based on the order of the addons being run). We can log it in that case. 

I'll update the RFC to account for this case.

**Comment by hjdivad** [comment]
@kratiahuja I'm concerned about adding features before we have an actual need for it.

Is there a reason to prefer the `from` feature over just erroring or overriding + warning for now until we have an actual case of two addons that could reasonably be used together but want to register import transforms with the same name?

Although I suppose it could be argued that guaranteeing addon order with override is a roundabout way of implementing `from`, I would see it more as a "soft error".

The tl;dr is my vote is to not support the case preemptively, but also to not have undefined semantics (hence the preference for hard or soft erroring)

**Comment by kratiahuja** [comment]
> Is there a reason to prefer the from feature over just erroring or overriding + warning for now until we have an actual case of two addons that could reasonably be used together but want to register import transforms with the same name?

Since the ordering is governed by the addons that define the transform, for an app that wants to use the the transform from an addon that has the same name as defined by the other addon becomes extremely cumbersome to do so. They have to ask the addon owner that defined the transform to change the name. I am fine with the default behavior being overridden + warning when `from` not used but I think it is important to provide a complete solution for such edge cases. 

The reason I am hesitant over erroring it is since apps that consume two addons (that provide the same transform name) are now left with the choice of choosing which addon to keep in their list in order to use the correct version/have their build passing. I think it is important to not introduce friction in such cases for apps.

**Comment by kratiahuja** [comment]
Updated per feedback from the June end ember-cli team meeting.

**Comment by kratiahuja** [comment]
@chadhietala Would love your feedback on this RFC wrt if this is going to be an issue with tree shaking. I am told you had some thoughts on `app.import` in general that was discussed in ember-cli core meeting in Portland.

**Comment by chadhietala** [comment]
I believe we have decided that anything that comes through `app.import` is treated as opaque and just concated into a bundle. I don't have issues with adding a transform to wrap the stuff going through `app.import` as long as the input from doing so is held separately from the files that would candidates for shaking.

**Comment by kratiahuja** [comment]
Sure sounds good. Thanks @chadhietala

**Comment by hjdivad** [comment]
We have discussed this during the weekly meetings and decided to move to FCP

**Comment by kratiahuja** [comment]
Implementation of RFC: https://github.com/ember-cli/ember-cli/pull/7245

**Comment by kratiahuja** [comment]
I changed the` importTransforms` API to accomodate `processOptions` functions as well. `processOptions` should be defined if you want to process an additional properties app developers provide along with the transformName. `importTranforms` can either define only the callback or callback and processOptions.

Therefore it can have a structure as when defining `processOptions`:
```js
importTransforms() {
  return {
     'amd': { 
        transform: function(tree, options) {},
        processOptions: function(assetPath, entry, options) {
            // do the processing
            return options;
       }
     }
  }
}
```

When the processOptions does not need to be defined, the API can be used as:
```js
importTransforms() {
  return {
     'amd': function(tree, options) {},
  }
}
```
I rarely see addons requiring to processOptions, therefore wanted to make sure we support both.

**Comment by rwjblue** [comment]
Awesome, thank you for updating on the implementation progress.

Considering that there is no new discussion here, and that implementation has been going well, I think we should merge.

**Comment by twokul** [comment]
🎉
